0001-typedb/02-dev/03-match.md

```diff
@@ -0,0 +1,755 @@
+
+To read data from a database use get query.
+
+<!-- test-ignore -->
+```typeql
+match 
+  $u isa user, has credential $cr; 
+get $cr;
+```
+
+The above example matches all instances of the `user` type and all its subtypes that have `credential` attribute. 
+Get clause filters the result so only the `credential` attributes will be returned.
+
+For more information on how to read data please see the [Reading data](05-read.md) page.
+
+#### Insert query
+
+To write new data to a database use insert query.
+
+<!-- test-ignore -->
+```typeql
+match
+  $p isa person, has full-name "Bob";
+insert
+  $p has email "bob@vaticle.com";
+```
+
+The above example matches all instances of the `person` type and all its subtypes that have `full-name` attribute 
+with value of `Bob`. Then it inserts the data that all matched instances of `person` have `email` attribute with 
+value of `bob@vaticle.com`.
+
+For more information on how to write data please see the [Writing data](04-write.md#insert-query) page.
+
+#### Delete query
+
+To delete data from a database use delete query.
+
+<!-- test-ignore -->
+```typeql
+match
+  $p isa person, has email "bob@vaticle.com";
+delete
+  $p isa person;
+```
+
+The above example matches all instances of the `person` type and all its subtypes that have `email` attribute 
+with value of `bob@vaticle.com`. Then it deletes the matched instances of `person`.
+
+For more information on how to delete data please see the [Writing data](04-write.md#delete-query) page.
+
+#### Update
+
+To update data in a database we use a combined `match-delete-insert` query.
+
+<!-- test-ignore -->
+```typeql
+match
+  $p isa person, has full-name "Masako Holley", has email $email;
+delete
+  $p has $email;
+insert
+  $p has email "m.holley@vaticle.com";
+```
+
+The above example matches all instances of the `person` type and all its subtypes that have `full-name` attribute 
+with value of `Masako Holley` and have `email` attribute (assigning the `$email` variable). It proceeds with 
+deleting the ownership of any `email` attributes that the matched `person` instances have. Finally, it inserts 
+the data that all matched instances of `person` have `email` attribute with value of `m.holley@vaticle.com`.
+
+For more information on how to update data please see the [Writing data](04-write.md#update) page.
+
+### Complex query example
+
+A `match` clause can only address data or types that already exist in a database.
+
+The **declarative** nature of the TypeQL means that if one of the `match` clause patterns can’t find any instances 
+in a database that might or might not lead to an empty response.
+
+For example:
+
+<!-- test-ignore -->
+```typeql
+match
+  $p isa person, has full-name $p-fname;
+  $o isa object, has path $o-path;
+  $a isa action, has action-name "view_file";
+  $ac(accessed-object: $o, valid-action: $a) isa access;
+  $pe(permitted-subject: $p, permitted-access: $ac) isa permission;
+  $p-fname = "Kevin Morrison";
+```
```

Comment:
I think, many examples in this file would benefit from being preceded by motivation, as in what are we trying to achieve. That short explanation could also inform the structure of the example query.

Using this as an example (rough sketch):

We would like to retrieve the list of objects that someone named "Kevin Morrison" has a `view_file` access permission to:

```typeql
match
  $p isa person, has full-name "Kevin Morrison";
  $pe(permitted-subject: $p, permitted-access: $ac) isa permission;
  $ac(accessed-object: $o, valid-action: $a) isa access;
  $o isa object, has path $o-path;
  $a isa action, has action-name "view_file";
```
1. Matches a `person` with `full-name` "Kevin Morrison".
2. Matches all `permission` relations (`$pe`) that relate `$p` (as `permitted-subject` role) to `$ac` (as `permitted-access` role).
3. ...

Reply:
This will be duplicating a lot of content that goes after the query - where we explain every statement in detail. And provide example of how it will work in a query.